feature: dune build @doc-json

emillon opened the pull request:
The `odoc` tool supports a `--as-json` flag. When used in `odoc html-generate`,
it will output `.html.json` files instead of `.html` files.

These files contain HTML fragments that can be used by external tools to generate HTML documents using a different pipeline.

This is exposed in Dune by defining a new `@doc-json` alias that works like `@doc` (it builds the documentation for public packages) but emits JSON files.

emillon committed:
Add Toplevel_index

Signed-off-by: Etienne Millon <me@emillon.org>

emillon committed:
Add JSON toplevel index

Signed-off-by: Etienne Millon <me@emillon.org>

Alizter commented:
I wonder if it would be worth making our own little json library in `otherlibs` that can be used by chrometrace, the bench and this.

emillon commented:
@jonludlam can you have a look at this? thanks!

emillon committed:
feature: dune build @doc-json

The `odoc` tool supports a `--as-json` flag. When used in `odoc
html-generate`, it will output `.html.json` files instead of `.html`
files.

These files contain HTML fragments that can be used by external tools to
generate HTML documents using a different pipeline.

This is exposed in Dune by defining a new `@doc-json` alias that works
like `@doc` (it builds the documentation for public packages) but emits
JSON files.

Signed-off-by: Etienne Millon <me@emillon.org>

emillon commented:
Yes, it's still expected to land on 3.11. But this isn't likely to break any existing code, so it can be included in a later alpha (or in the final release).

jonludlam reviewed (approved):
Looks good!

emillon commented on src/dune_rules/odoc.ml:
```diff
@@ -111,9 +112,50 @@ module Paths = struct
+  let extension = function
+    | Html -> ".html"
+    | Json -> ".html.json"
+  ;;
```
that comes from a change in our ocamlformat config. we now use the JS profile. one advantage of double semis is better location for error messages a `in` is missing.

jonludlam replied:
aha! OK, that makes sense.

emillon commented on src/dune_rules/odoc.ml:
```diff
@@ -448,10 +506,48 @@ let setup_toplevel_index_rule sctx =
+      ]
+  ;;
+
+  (** This format is public API. *)
```
There's no formal schema for this, but this is printed in full in a test to avoid changing accidentally. Would you prefer a schema in this docstring?

Re #7840, I think that the question is what are you doing on the HTML side of things? If the toplevel index is modified, I think that the JSON one should too. But we don't have important versioning constraints here (the schema of the output depends on the version of dune itself).

emillon replied:
OK. I see ne problem then, this JSON file is going to disappear in a few releases then.

emillon committed:
Inline html_alias

Signed-off-by: Etienne Millon <me@emillon.org>
